librz/include/rz_lib.h

```diff
@@ -31,95 +29,111 @@ RZ_LIB_VERSION_HEADER(rz_lib);
 #define RZ_LIB_EXT "so"
 #endif
 
-/* store list of loaded plugins */
+/**
+ * \brief Identify the Plugin type
+ */
+typedef enum {
+	RZ_LIB_TYPE_IO, ///< io layer
+	RZ_LIB_TYPE_DBG, ///< debugger
+	RZ_LIB_TYPE_LANG, ///< language
+	RZ_LIB_TYPE_ASM, ///< assembler
+	RZ_LIB_TYPE_ANALYSIS, ///< analysis
+	RZ_LIB_TYPE_PARSE, ///< parsers
+	RZ_LIB_TYPE_BIN, ///< bin headers
+	RZ_LIB_TYPE_BIN_XTR, ///< bin extractors
+	RZ_LIB_TYPE_BIN_LDR, ///< DEPRECATED
+	RZ_LIB_TYPE_BP, ///< breakpoint
+	RZ_LIB_TYPE_SYSCALL, ///< DEPRECATED
+	RZ_LIB_TYPE_FASTCALL, ///< DEPRECATED
+	RZ_LIB_TYPE_CRYPTO, ///< cryptography
+	RZ_LIB_TYPE_HASH, ///< hashes / message digests
+	RZ_LIB_TYPE_CORE, ///< RzCore commands
+	RZ_LIB_TYPE_EGG, ///< rz_egg plugin
+	RZ_LIB_TYPE_DEMANGLER, ///< demanglers
+	RZ_LIB_TYPE_UNKNOWN
+} RzLibType;
+
+/**
+ * \brief Represent a single Plugin
+ */
 typedef struct rz_lib_plugin_t {
-	int type;
-	char *file;
-	void *data; /* user pointer */
-	struct rz_lib_handler_t *handler;
-	void *dl_handler; // DL HANDLER
-	char *author;
-	char *version;
+	RzLibType type; ///< Type of the plugin
+	char *file; ///< File path
+	void *data; ///< User pointer
+	struct rz_lib_handler_t *handler; ///< Handler that handles this plugin
+	char *author; ///< Author of the plugin
+	char *version; ///< Version of the plugin
 	void (*free)(void *data);
 } RzLibPlugin;
 
-/* store list of initialized plugin handlers */
+typedef bool (*RzLibCallback)(RzLibPlugin *, void *, void *);
```

Comment:
If the constructor has to return something, like a pointer, how can it? for the destructor this makes sense, but not for the constructor

Reply:
Where do you see the need to return anything here? I think the purpose is usually just to call something like `rz_crypto_plugin_add` and make sure RzCrypto is aware of the new plugin.

Reply:
the constructor should allow to set a user datastructure and the destructure to destroy it (we have the user pointer on purpose. but using this function declaration, it does not allow to set anything.

Reply:
The constructor can manipulate the `RzLibPlugin.data` pointer, which is usually `RzCryptoPlugin`, `RzAsmPlugin`, etc. and there it can manipulate everything. Do you have concrete examples of what would be useful here?